src/software/ai/hl/stp/evaluation/robot.cpp

```diff
@@ -16,31 +18,86 @@ bool Evaluation::robotOrientationWithinAngleThresholdOfTarget(const Point positi
     return diff_orientation < threshold;
 }
 
-bool Evaluation::robotHasPossession(Ball ball, Robot robot, Timestamp timestamp)
+bool Evaluation::robotHasPossession(const Ball& ball, const Robot& robot,
+                                    Timestamp timestamp)
 {
-    int ball_index  = 0;
-    int robot_index = 0;
-    if (timestamp.getMilliseconds() != 0)
+    Point robot_pos_at_time;
+    Angle robot_ori_at_time;
+    Point ball_pos_at_time;
+
+
+    if (robot.getHistoryIndexFromTimestamp(timestamp))
+    {
+        robot_pos_at_time = robot.getPreviousPositions().at(
+            *robot.getHistoryIndexFromTimestamp(timestamp));
+        robot_ori_at_time = robot.getPreviousOrientations().at(
+            *robot.getHistoryIndexFromTimestamp(timestamp));
+    }
+    else
     {
-        // Get the indices for the state information at the desired timestamp.
-        auto ball_index_result  = ball.getHistoryIndexFromTimestamp(timestamp);
-        auto robot_index_result = robot.getHistoryIndexFromTimestamp(timestamp);
-        if (!ball_index_result || !robot_index_result)
-            return false;
+        robot_pos_at_time = robot.position();
+        robot_ori_at_time = robot.orientation();
```

Comment:
Looks like we're not testing this `else` case and the other ones below. Happy to write tests for these, just lemme know.

Reply:
I'll let you handle those :+1:

Reply:
Created PR to your branch: https://github.com/henryymliu/UBC-Thunderbots-Software/pull/3